### PR TITLE
[FLINK-14825][state-processor-api][docs] Rework state processor api documentation

### PR DESCRIPTION
**Pull request by sjwiesman**
## What is the purpose of the change

Update the state processor api documentation based on user feedback to clarify usage and examples. The introduction is taken from the state proc blog post on the flink blog[1]. 

[1] https://flink.apache.org/feature/2019/09/13/state-processor-api.html

### CODE REVIEW

**Comment by sjwiesman** [comment]
cc @NicoK

File: docs/dev/libs/state_processor_api.md [hunk 1]
```diff
@@ -23,166 +23,99 @@ specific language governing permissions and limitations
...
 {% highlight scala %}
 val bEnv      = ExecutionEnvironment.getExecutionEnvironment()
-val savepoint = Savepoint.load(bEnv, "hdfs://path/", new RocksDBStateBackend())
+val savepoint = Savepoint.load(bEnv, "hdfs://path/", new MemoryStateBackend())
 {% endhighlight %}
 </div>
 </div>
 
-When reading operator state, simply specify the operator uid, state name, and type information.
+### Operator State
```
Comment (qinjunjerry) [inline docs/dev/libs/state_processor_api.md]: I think adding the following generalized sentence here before expanding to different kind of operator states is helpful, especially for understanding those parameters given in the followed examples:

>When reading operator state, users specify the operator uid, the state name, and the type information.

File: docs/dev/libs/state_processor_api.md [hunk 2]
```diff
@@ -229,6 +206,8 @@ val listState = savepoint.readListState(
 </div>
 </div>
 
+### Keyed State
```
Comment (qinjunjerry) [inline docs/dev/libs/state_processor_api.md]: Similar here:
> When reading a keyed state, users specify the operator id and a KeyedStateReaderFunction. Using KeyedStateReaderFunction allows reading...

File: docs/dev/libs/state_processor_api.md [hunk 3]
```diff
@@ -333,20 +318,165 @@ val keyedState = savepoint.readKeyedState("my-uid", new ReaderFunction());
...
-{% panel **Note:** When using a `KeyedStateReaderFunction` all state descriptors must be registered eagerly inside of open. Any attempt to call `RuntimeContext#getState`, `RuntimeContext#getListState`, or `RuntimeContext#getMapState` will result in a `RuntimeException`. %}
+Along with reading registered state values, each key has access to a `Context` with metadata such as registered event time and processing time timers.
+
+{% panel **Note:** When using a `KeyedStateReaderFunction` all state descriptors must be registered eagerly inside of open. Any attempt to call a `RuntimeContext#get*State` will result in a `RuntimeException`. %}
 
 ## Writing New Savepoints
 
-State writers are based around the abstraction of `Savepoint`, where one `Savepoint` may have many operators and the state for any particular operator is created using a `BootstrapTransformation`.
+`Savepoint`'s may also be written, which allows such use cases as bootstrapping state based on historical data.
+Each savepoint is made up of many `BootstrapTransformation`'s (explained below), which define the state for an individual operator.
```
Comment (qinjunjerry) [inline docs/dev/libs/state_processor_api.md]: Maybe this?
>... Each savepoint is made up of some `BootstrapTransformation`'s (explained below), each of which defines the state for an individual operator.

**Comment by qinjunjerry** [comment]
Other than the comments above, it was a good reading: more clear with the pictures :)

**Comment by sjwiesman** [comment]
@qinjunjerry Thanks for taking a look. I believe I have addressed all your comments if you'd like to take another pass.

**Comment by qinjunjerry** [comment]
Thanks for the update. All looks good to me :)

File: docs/dev/libs/state_processor_api.md [hunk 4]
```diff
@@ -333,20 +325,165 @@ val keyedState = savepoint.readKeyedState("my-uid", new ReaderFunction());
...
+</div>
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+case class CurrencyRate(currency: String, rate: Double)
+
+class CurrencyBootstrapFunction extends BroadcastStateBootstrapFunction[CurrencyRate] {
+
+    @throws[Exception]
+    override processElement(value: CurrencyRate, ctx: Context): Unit = {
+        ctx.getBroadcastState(descriptor).put(value.currency, value.rate)
```
Comment (NicoK) [inline docs/dev/libs/state_processor_api.md]: Can you add the `descriptor` definition somewhere?

File: docs/dev/libs/state_processor_api.md [hunk 5]
```diff
@@ -333,20 +325,165 @@ val keyedState = savepoint.readKeyedState("my-uid", new ReaderFunction());
...
+    public String currency;
+
+    public Double rate;
+}
+
+public class CurrencyBootstrapFunction extends BroadcastStateBootstrapFunction<CurrencyRate> {
+
+    @Override
+    public void processElement(CurrencyRate value, Context ctx) throws Exception {
+        ctx.getBroadcastState(descriptor).put(value.currency, value.rate);
```
Comment (NicoK) [inline docs/dev/libs/state_processor_api.md]: Can you add the `descriptor` definition somewhere?

File: docs/dev/libs/state_processor_api.md [hunk 6]
```diff
@@ -413,45 +551,25 @@ The `KeyedStateBootstrapFunction` supports setting event time and processing tim
 The timers will not fire inside the bootstrap function and only become active once restored within a `DataStream` application.
 If a processing time timer is set but the state is not restored until after that time has passed, the timer will fire immediatly upon start.
```
Comment (NicoK) [inline docs/dev/libs/state_processor_api.md]:
```suggestion
If a processing time timer is set but the state is not restored until after that time has passed, the timer will fire immediately upon start.
```

File: docs/dev/libs/state_processor_api.md [hunk 7]
```diff
@@ -413,45 +551,25 @@ The `KeyedStateBootstrapFunction` supports setting event time and processing tim
...
-Savepoint
-    .create(backend, 128)
-    .withOperator("uid1", transformation1)
-    .withOperator("uid2", transformation2)
-    .write(savepointPath)
-{% endhighlight %}
-</div>
-</div>
-		
 Besides creating a savepoint from scratch, you can base on off an existing savepoint such as when bootstrapping a single new operator for an existing job.
```
Comment (NicoK) [inline docs/dev/libs/state_processor_api.md]:
```suggestion
Besides creating a savepoint from scratch, you can base one off an existing savepoint such as when bootstrapping a single new operator for an existing job.
```

File: docs/dev/libs/state_processor_api.md [hunk 8]
```diff
@@ -413,45 +551,25 @@ The `KeyedStateBootstrapFunction` supports setting event time and processing tim
...
 {% endhighlight %}
 </div>
 <div data-lang="scala" markdown="1">
 {% highlight scala %}
 Savepoint
-    .load(backend, oldPath)
+    .load(new MemoryStateBackend(), oldPath)
     .withOperator("uid", transformation)
     .write(newPath)
 {% endhighlight %}
```
Comment (NicoK) [inline docs/dev/libs/state_processor_api.md]: About potential corruption below: I guess, you could delete the new savepoint though. Can you rephrase that?

File: docs/dev/libs/state_processor_api.md [hunk 9]
```diff
@@ -23,166 +23,103 @@ specific language governing permissions and limitations
...
+## Mapping Application State to DataSets 
 
-To understand how to best interact with savepoints in a batch context it is important to have a clear mental model of how the data in Flink state relates to a traditional relational database.
+The State Processor API maps the state of a streaming application to one or more data sets that can be separately processed.
+In order to be able to use the API, you need to understand how this mapping works.
 
-A database can be thought of as one or more namespaces, each containing a collection of tables.
-Those tables in turn contain columns whose values have some intrinsic relationship between them, such as being scoped under the same key.
+But let's first have a look at what a stateful Flink job looks like.
+A Flink job is composed of operators, typically one or more source operators, a few operators for the actual processing, and one or more sink operators.
```
[the remark this answers is not in the document]
Comment (sjwiesman) [reply docs/dev/libs/state_processor_api.md]: I think it should actually be a ';' but yes

File: docs/dev/libs/state_processor_api.md (hunk 8 above)
Comment (sjwiesman) [reply docs/dev/libs/state_processor_api.md]: No you cannot delete the new savepoint. It would delete the shared operators.

File: docs/dev/libs/state_processor_api.md [hunk 10]
```diff
@@ -23,166 +23,103 @@ specific language governing permissions and limitations
...
 ExecutionEnvironment bEnv   = ExecutionEnvironment.getExecutionEnvironment();
-ExistingSavepoint savepoint = Savepoint.load(bEnv, "hdfs://path/", new RocksDBStateBackend());
+ExistingSavepoint savepoint = Savepoint.load(bEnv, "hdfs://path/", new MemoryStateBackend());
 {% endhighlight %}
 </div>
 <div data-lang="scala" markdown="1">
 {% highlight scala %}
 val bEnv      = ExecutionEnvironment.getExecutionEnvironment()
-val savepoint = Savepoint.load(bEnv, "hdfs://path/", new RocksDBStateBackend())
+val savepoint = Savepoint.load(bEnv, "hdfs://path/", new MemoryStateBackend())
```
[the remark this answers is not in the document]
Comment (sjwiesman) [reply docs/dev/libs/state_processor_api.md]: I've really grown to dislike scala

File: docs/dev/libs/state_processor_api.md [hunk 11]
```diff
@@ -276,12 +264,13 @@ Then it can read by defining an output type and corresponding KeyedStateReaderFu
 <div class="codetabs" markdown="1">
 <div data-lang="java" markdown="1">
 {% highlight java %}
-class KeyedState {
-  Integer key;
-  Integer value;
+public class KeyedState {
```
[the remark this answers is not in the document]
Comment (sjwiesman) [reply docs/dev/libs/state_processor_api.md]: 👍

**Comment by sjwiesman** [comment]
@NicoK I believe I have addressed all your comments except: 
1) I did not update the picture to change `v1`, `v2` to another letter. I did not create the graphic so I will need to find the person who did. I'm not sure how important you think that is. 

2) Async I/O functions are not stateful so I'm not sure there's anything to say there

3) I'd prefer to just add window documentation in the branch with that feature

File: docs/dev/libs/state_processor_api.md [hunk 12]
```diff
@@ -239,114 +225,302 @@ public class StatefulFunctionWithTime extends KeyedProcessFunction<Integer, Inte
...
+Savepoint
+    .create(new MemoryStateBackend(), maxParallelism)
+    .withOperator("uid1", transformation1)
+    .withOperator("uid2", transformation2)
+    .write(savepointPath)
+{% endhighlight %}
+</div>
+</div>
+
+The [UIDs]({{ site.baseurl}}/ops/state/savepoints.html#assigning-operator-ids) associated with each operator much match one to one with the UIDs assigned to the operators in your `DataStream` application; these are how Flink knows what state maps to which operator.
```
Comment (qinjunjerry) [inline docs/dev/libs/state_processor_api.md]: Typo: ... each operator **much** match ... -> ... each operator much **must** ...

File: docs/dev/libs/state_processor_api.md (hunk 8 above)
Comment (qinjunjerry) [reply docs/dev/libs/state_processor_api.md]: I believe there are three parameters for Savepoint.load(): ExecutionEnvironment, a path and StateBackend. Could you double check? To me, the Savepoint.load() call at the beginning of the "Reading State" section is a good one.

File: docs/dev/libs/state_processor_api.md (hunk 12 above)
Comment (sjwiesman) [reply docs/dev/libs/state_processor_api.md]: Nice catch!

**Comment by sjwiesman** [comment]
cc @NicoK

File: docs/dev/libs/state_processor_api.md (hunk 8 above)
Comment (NicoK) [reply docs/dev/libs/state_processor_api.md]: Good catch, @qinjunjerry 

@sjwiesman: what I meant is that you can always remove the new savepoint without affecting the old one because the old one is not touched, is it?

Comment (NicoK) [reply docs/dev/libs/state_processor_api.md]: ok, let's stay cautious here - it depends how you delete the new savepoint: if Flink (in the future) would expose a savepoint delete method, this would then only look at the metadata file and also delete files from the old one, so we keep the sentence as it is
